### ISSUE TITLE
Live plots do not automatically update on Codespaces

### ISSUE THREAD

**Issue by shcheklein** [iterative/vscode-dvc#3264]
Looks like something weird is happening with our watchers on Codedspaces. Local dev version runs fine for me. 

## How to reproduce

I've tried both - GPU and non-GPU setup. Two repositories: [Get Started](https://github.com/iterative/example-get-started-experiments) and [Hackathon](https://github.com/shcheklein/hackathon).

I'm getting only the first dot for the first epoch and then only at the very end it updates everything:

https://user-images.githubusercontent.com/3659196/218223477-c201c726-d819-4e23-bef5-13e7486cbdb5.mov

https://user-images.githubusercontent.com/3659196/218223521-e099b00e-6e8f-4b2a-a043-61f5f35c2e20.mov

## Environment

Extension version: 0.6.4
DVC Version:

```
DVC version: 2.44.0 (deb)
-------------------------
Platform: Python 3.10.8 on Linux-5.4.0-1101-azure-x86_64-with-glibc2.31
Subprojects:

Supports:
        azure (adlfs = 2023.1.0, knack = 0.10.1, azure-identity = 1.12.0),
        gdrive (pydrive2 = 1.15.0),
        gs (gcsfs = 2023.1.0),
        hdfs (fsspec = 2023.1.0, pyarrow = 11.0.0),
        http (aiohttp = 3.8.3, aiohttp-retry = 2.8.3),
        https (aiohttp = 3.8.3, aiohttp-retry = 2.8.3),
        oss (ossfs = 2021.8.0),
        s3 (s3fs = 2023.1.0, boto3 = 1.24.59),
        ssh (sshfs = 2023.1.0),
        webdav (webdav4 = 0.9.8),
        webdavs (webdav4 = 0.9.8),
        webhdfs (fsspec = 2023.1.0)
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/loop0
Caches: local
Remotes: https
Workspace directory: ext4 on /dev/loop0
Repo: dvc, git
```

`dvclive==2.0.2`

**Comment by shcheklein** [comment]
Making it a `p1` priority since it affects us demoing this to customers. @alex000kim @daavoo do you also see this behavior?

**Comment by mattseddon** [comment]
I can recreate the behaviour in `example-get-started` but not `magnetic-tiles-defects`.

Looks like there is at least one issue related to images:

<img width="487" alt="image" src="https://user-images.githubusercontent.com/37993418/218344004-5cd0d427-1a82-483b-b862-1551cb59b2d1.png">

I think it might be a timing issue between `plots diff` writing the file to the `tmp` directory and the message being sent to the webview/the webview trying to load the file. Will come back to this once I've got to a breakpoint in the Studio integration work.

**Comment by alex000kim** [comment]
> @alex000kim @daavoo do you also see this behavior?

Yes, I saw this today in https://github.com/iterative/example-get-started-experiments
My workaround when demoing is to keep pressing the "Refresh" button.

**Comment by mattseddon** [comment]
`0.6.8` didn't fix the issue. I'll keep investigating.

**Comment by mattseddon** [comment]
@alex000kim @shcheklein do either of you know the magic incantation to run `example-get-started-experiments` on M1? I am running into this error:

```
loc("mps_scatter_along_axis"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/9e200cfa-7d96-11ed-886f-a23c4f261b56/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":350:0)): error: invalid input tensor shapes, indices shape and updates shape must be equal
/AppleInternal/Library/BuildRoots/9e200cfa-7d96-11ed-886f-a23c4f261b56/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphExecutable.mm:1641: failed assertion `Error: MLIR pass manager failed'
ERROR: failed to reproduce 'train': failed to run: python src/train.py, exited with -6
```

Have spent some time trying to find an answer but have been unable to make it work.

**Comment by mattseddon** [comment]
There seems to be an issue with watching heavily-nested folder structures on Codespaces. Below is a demo of the depth where events stop getting generated:

https://user-images.githubusercontent.com/37993418/219987038-f5ab8b90-5bac-410c-8957-fa7b70271257.mov

Looking at options now.

**Comment by mattseddon** [comment]
Not a viable option but partly confirms the above theory. In this demo I have patched the `DVCLiveCallback` to touch the main `dvc.yaml` in the `after_epoch` function:


https://user-images.githubusercontent.com/37993418/219994251-d76b6925-30d3-43ec-96fe-40ad8b8f4756.mov

**Comment by daavoo** [comment]
> Not a viable option but partly confirms the above theory. In this demo I have patched the DVCLiveCallback to touch the main dvc.yaml in the after_epoch function:

`dvclive/metrics.json` always gets updated on `next_step`

**Comment by mattseddon** [comment]
Thanks @daavoo as a patch we can rely on `metrics.json` being updated 🙏🏻.

I can think of a couple of cases when this won't work:

- `metrics.json` is still too deep/events are still dropped. E.g `DVCLive` dir is set to `model/results/stage/train`. 
- Users not working with `DVCLive`.

**Comment by mattseddon** [comment]
With patch in place: 

https://user-images.githubusercontent.com/37993418/220210459-3a50329e-fdec-4de0-beec-d36686b2fcb5.mov

https://user-images.githubusercontent.com/37993418/220210582-8075a619-3901-4900-a367-6ddb25ee84c6.mov

**Comment by daavoo** [comment]
> metrics.json is named something else.
> Users not working with DVCLive.

These 2 are the same. DVCLive has `metrics.json` hardcoded as name, users can't change it

**Comment by mattseddon** [comment]
The second patch (#3321) watches all metrics files for updates (i.e all metrics files from the `exp show` `workspace` record). I think this is good enough to close this for now. We can reopen if (probably when) the issue reappears.

**Comment by mattseddon** [comment]
> There seems to be an issue with watching heavily-nested folder structures on Codespaces. Below is a demo of the depth where events stop getting generated:

FWIW I have checked to see if this was a Linux issue by creating the devcontainer from `example-get-started-experiments` with locally with Docker. Training performance was TERRIBLE but I could not recreate the issue. All events were created as expect.

**Comment by shcheklein** [comment]
Thanks @mattseddon for the update, it makes sense.

I'm curious what is so special about Codespaces then ... I guess we would need to ask them. It's fine to this in background as we discussed.

**Comment by mattseddon** [comment]
Raised https://github.com/microsoft/vscode-discussions/discussions/486, will see if it goes anywhere.

**Comment by mattseddon** [comment]
Raised https://github.com/community/community/discussions/48658 as well.

**Comment by mattseddon** [comment]
For the record here. Got a response in `community/community`:

![image](https://user-images.githubusercontent.com/37993418/222287345-68a6ec43-b6a3-40d5-8cf0-85f408dcbcf7.png)

**Comment by mattseddon** [comment]
Now raised https://github.com/microsoft/vscode/issues/176327
